README.md
# shenshang1


HISTORY.rst
=======
History
=======

0.1.0 (2019-08-11)
------------------

* First release on PyPI.


tests/test_cli.py
from click.testing import CliRunner

from shenshang.cli import main


def test_binary():
  runner = CliRunner()
  result = runner.invoke(main, ['binary', '-h'])
  assert result.exit_code == 0
  # assert result.output == 'Hello Peter!\n'


tests/test_util.py
import pytest
import pandas as pd
import numpy as np

from shenshang.util import (
    fetch_tri, convert_format, compare_matrices, argsort_arrays, combine)


@pytest.mark.parametrize(
    'square, lower, exp1, exp2, exp3',
    [(np.array([[1.0, 0.2],
                [0.1, 1.0]]),
      True,
      np.array([0.1]), np.array([1]), np.array([0])),
     (np.array([[1.0, 0.2],
                [0.1, 1.0]]),
      False,
      np.array([0.2]), np.array([0]), np.array([1])),
     (np.array([[np.nan, np.nan, np.nan],
                [np.nan, np.nan, 0.6],
                [0.7,    0.8,    np.nan]]),
      True,
      np.array([np.nan, 0.7, 0.8]), np.array([1, 2, 2]), np.array([0, 0, 1])),
     (np.array([[np.nan, np.nan, 0.6],
                [np.nan, np.nan, np.nan],
                [0.7,    0.8,    np.nan]]),
      False,
      np.array([np.nan, 0.6, np.nan]), np.array([0, 0, 1]), np.array([1, 2, 2]))])
def test_fetch_tri(square, lower, exp1, exp2, exp3):
    p, idx1, idx2 = fetch_tri(square, lower=lower)
    np.testing.assert_array_equal(p, exp1)
    assert np.all(idx1 == exp2)
    assert np.all(idx2 == exp3)


@pytest.mark.parametrize(
    'stat, pval, cutoff, exp',
    [(pd.DataFrame([[np.nan, 9.9],
                    [np.nan, np.nan]]),
      pd.DataFrame([[np.nan, 0.099],
                    [0.91,   np.nan]]),
      0.1,
      pd.DataFrame([[0, 1, 9.9, 0.099, '+']],
                    columns=('feature1', 'feature2', 'stat', 'pval', 'type'))),
     (pd.DataFrame([[np.nan, np.nan, 6.0],
                    [np.nan, np.nan, 7.0],
                    [np.nan, np.nan, np.nan]],
                   index=['OTU1', 'OTU2', 'OTU3'],
                   columns=['OTU1', 'OTU2', 'OTU3']),
      pd.DataFrame([[np.nan, np.nan, 0.001],
                    [np.nan, np.nan, np.nan],
                    [0.7,    0.008,  np.nan]],
                   index=['OTU1', 'OTU2', 'OTU3'],
                   columns=['OTU1', 'OTU2', 'OTU3']),
      0.01,
      pd.DataFrame([['OTU1', 'OTU3', 6.0, 0.001, '+'],
                    ['OTU3', 'OTU2', 7.0, 0.008, '-']],
                    columns=('feature1', 'feature2', 'stat', 'pval', 'type')))])
def test_convert_format(stat, pval, cutoff, exp):
    df = convert_format(stat, pval, cutoff=cutoff)
    pd.util.testing.assert_frame_equal(exp, df)


@pytest.mark.parametrize(
    'pvals, sigs, cutoff, exp',
    [((pd.DataFrame([[np.nan, 0.01],
                     [0.02,   np.nan]]),
       pd.DataFrame([[np.nan, 0.03],
                     [0.03,   np.nan]]),
       pd.DataFrame([[np.nan, 0.05],
                     [0.06,   np.nan]])),
      [True, True, False],
      0.04,
      pd.DataFrame([[0, 1, '+', 0.01, 0.03, 0.05],
                    [1, 0, '-', 0.02, 0.03, 0.06]])),
     ((pd.DataFrame([[np.nan, 0.01],
                     [0.02,   np.nan]],
                    index=['f1', 'f2'],
                    columns=['f1', 'f2']),
       pd.DataFrame([[np.nan, 0.05],
                     [0.06,   np.nan]],
                    index=['f1', 'f2'],
                    columns=['f1', 'f2'])),
      [False, True],
      0.01,
      pd.DataFrame(columns=[0, 1, 2, 3, 4],
                   index=pd.RangeIndex(0),
                   dtype=float)),
     ((pd.DataFrame([[np.nan, 0.03],
                     [0.04,   np.nan]],
                    index=['f1', 'f2'],
                    columns=['f1', 'f2']),
       pd.DataFrame([[np.nan, 0.01],
                     [0.02,   np.nan]],
                    index=['f1', 'f2'],
                    columns=['f1', 'f2']),
       pd.DataFrame([[np.nan, 0.05],
                     [0.06,   np.nan]],
                    index=['f1', 'f2'],
                    columns=['f1', 'f2'])),
      [False, True, False],
      0.03,
      pd.DataFrame([['f1', 'f2', '+', 0.03, 0.01, 0.05],
                    ['f2', 'f1', '-', 0.04, 0.02, 0.06]]))])
def test_compare_matrices(pvals, sigs, cutoff, exp):
    df = compare_matrices(pvals, sigs, cutoff=cutoff)
    pd.util.testing.assert_frame_equal(df, exp)


@pytest.mark.parametrize(
    'arrs, ascending, exps',
    [((np.array([1, 2, 1, 3, 1, 4, 1, 5, 1, 6]),
       np.array([6, 5, 3, 4, 4, 9, 7, 1, 8, 9])),
      True,
      (np.array([1, 1, 1, 1, 1, 2, 3, 4, 5, 6]),
       np.array([3, 4, 6, 7, 8, 5, 4, 9, 1, 9]))),
     ((np.array([1, 6, 0, 1, 3, 1, 4, 1, 5, 1, 2, 0]),
       np.array([6, 5, 2, 3, 4, 4, 9, 7, 1, 8, 9, 7])),
      [True, False],
      (np.array([0, 0, 1, 1, 1, 1, 1, 2, 3, 4, 5, 6]),
       np.array([7, 2, 8, 7, 6, 4, 3, 9, 4, 9, 1, 5]))),
     ((np.array([0, 0, 0, 0, 0, 0, 0, 0, 0, 5, 1, 0, 0, 0, 0, 5, 0, 0, 0]),
       np.array([36,8, 20,16,36,3, 10,2, 1, 25,11,0, 2, 0, 6, 4, 10,38,4])),
      [True, False],
      (np.array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 5, 5]),
       np.array([38,36,36,20,16,10,10,8, 6, 4, 3, 2, 2, 1, 0, 0, 11,25,4]))),
     # if the input is a single array:
     ((np.array([3, 1, 2]),),
      False,
      (np.array([3, 2, 1]),))])
def test_argsort_arrays(arrs, ascending, exps):
    idx = argsort_arrays(arrs, ascending)
    for arr, exp in zip(arrs, exps):
        assert np.all(exp == arr[idx])
    # from pprint import pprint
    # pprint(y[idx])
    # pprint(x[idx])


def test_combine():
    ids = ['o1', 'o2']
    pval1 = pd.DataFrame([[np.nan, 0.99],
                          [0.01,   np.nan]],
                         columns=ids,
                         index=ids)
    stat1 = pd.DataFrame([[np.nan, 99],
                          [np.nan,   np.nan]],
                         columns=ids,
                         index=ids)
    pval2 = pd.DataFrame([[np.nan, 0.9],
                          [0.1,   np.nan]],
                         columns=ids,
                         index=ids)
    stat2 = pd.DataFrame([[np.nan, 9],
                          [np.nan,   np.nan]],
                         columns=ids,
                         index=ids)
    pval3 = pd.DataFrame([[np.nan, 0.89],
                          [0.08,   np.nan]],
                         columns=ids,
                         index=ids)
    stat3 = pd.DataFrame([[np.nan, 8.9],
                          [np.nan, np.nan]],
                         columns=ids,
                         index=ids)
    obs = combine(cutoff=1, binary=(stat1, pval1), rank=(stat2, pval2), logratio=(stat3, pval3))
    exp = {'binary_stat': [99.0, 99.0],
           'binary_pval': [0.99, 0.01],
           'binary_type': ['+', '-'],
           'rank_stat': [9.0, 9.0],
           'rank_pval': [0.9, 0.1],
           'rank_type': ['+', '-'],
           'logratio_stat': [8.9, 8.9],
           'logratio_pval': [0.89, 0.08],
           'logratio_type': ['+', '-']}

    assert exp == obs.to_dict('list')


tests/test_cooccur.py
from contextlib import ExitStack as does_not_raise

import numpy as np
from scipy.stats import pearsonr
import pytest

from shenshang.cooccur import cooccur_binary, cooccur_logratio, cooccur_rank


# fixture is a bit like setUp in unittest.TestCase
# this fixture causes the parametrize_xy only invoked once for the current module.
@pytest.fixture(scope="module")
def parametrize_xy(request):
    arrays = dict(a=np.zeros(100),
                  b=np.ones(100),
                  # c: --========
                  # d: ---=======
                  # e: ==--------
                  # f: =====-----
                  # g: =======---
                  c=np.array([1] * 2 + [5] * 8),
                  d=np.array([1] * 3 + [2] * 7),
                  e=np.array([9] * 2 + [1] * 8),
                  f=np.array([7] * 5 + [1] * 5),
                  g=np.array([7] * 7 + [1] * 3),

                  h=np.arange(10),
                  i=np.arange(10, 0, -1),
                  j=np.array([0, 1, 20, 30, 40, 50, 60, 70, 80, 90]))

    return [arrays[var_name] for var_name in request.param]


@pytest.mark.parametrize('parametrize_xy',
                         [(('e', 'd')),
                          (('e', 'c')),
                          (('g', 'e')),
                          (('g', 'f')),
                          (('c', 'f')),
                          (('c', 'g'))],
                         indirect=['parametrize_xy'])
def test_cooccur_rank(parametrize_xy):
    x, y = parametrize_xy
    o, p, n, dist = cooccur_rank(x, y, 1000, seed=9)
    if o < 0:
        p = n
    r, pval = pearsonr(x, y)
    print(o, p)
    print(r, pval)


@pytest.mark.parametrize('parametrize_xy',
                         [('a', 'a'),
                          ('a', 'b'),
                          ('a', 'c')],
                         indirect=['parametrize_xy'])
def test_cooccur_binary_raise(parametrize_xy):
    x, y = parametrize_xy
    with pytest.raises(ValueError, match='x or y or both are absent or present in all samples'):
        cooccur_binary(x, y, cutoff=1)


@pytest.mark.parametrize('parametrize_xy, real, pval_pos, pval_neg',
                         [(('e', 'd'), (0+1)/(9+1),  0.9610, 0.0399),
                          (('e', 'c'), (0+1)/(10+1), 0.9900, 0.0109),
                          (('g', 'e'), (2+1)/(7+1),  0.2338, 0.7672),
                          (('g', 'f'), (5+1)/(7+1),  0.0430, 0.9580),
                          (('c', 'f'), (3+1)/(10+1), 0.8891, 0.1118),
                          (('c', 'g'), (5+1)/(10+1), 0.7702, 0.2307)],
                         indirect=['parametrize_xy'])
def test_cooccur_binary(parametrize_xy, real, pval_pos, pval_neg):
    x, y = parametrize_xy
    o, p, n, dist = cooccur_binary(x, y, cutoff=1, seed=9)
    assert o == pytest.approx(real, abs=0.0001)
    assert p == pytest.approx(pval_pos, abs=0.0001)
    assert n == pytest.approx(pval_neg, abs=0.0001)


shenshang/cli.py
# -*- coding: utf-8 -*-

"""Console script for shenshang."""
import sys
import os
from functools import partial
from logging import getLogger

import click
import pandas as pd

from .cooccur import compute_cooccur
from .util import convert_format


click.option = partial(click.option, show_default=True)


def decorator_composer(*decorators):
    def deco(f):
        for dec in reversed(decorators):
            f = dec(f)
        return f
    return deco


_common_parameters = decorator_composer(
    click.option('-i', '--input', type=click.Path(exists=True), required=True, help='input sample-by-feature table'),
    click.option('-f', '--format', type=click.Choice(['biom', 'tsv']), default='biom', help='input table format. For the tsv table, the features are in row and samples are in column; columns are separated by TABs; the 1st column must be feature IDs and the 1st row must be sample IDs; lines starting with `#` will be ignored.'),
    click.option('-o', '--output', type=click.STRING, required=True, help='output file prefix. There are 2 output files: 1. co-occur matrix stores the statistics/correlation coefficients; 2. p-value matrix stores the p-values.'),
    click.option('--force', is_flag=True, help='overwrite output file'),
    click.option('-p', '--progress', is_flag=True, help='show progress bar'),
    click.option('-v', '--verbose', count=True, help='Verbosity. Use multiple `v` (eg -vv) to increase verbosity'))


@click.group(context_settings=dict(
    help_option_names=['-h', '--help'],
    # allow case insensitivity for the (sub)commands and their options
    token_normalize_func=lambda x: x.lower()))
def main():
    """Compute co-occurrence or mutual exclusivity between features."""


@main.command()
@_common_parameters
@click.option('--cutoff', type=click.INT, default=2, help='cutoff to determine presence and absence')
@click.option('--psudo', type=click.IntRange(1), default=1, help='psudo number added to nominator and denominator for overlap computation')
@click.option('--perms', type=click.INT, default=1000, help='the number of permutations to compute p-value')
@click.option('-c', '--cpus', type=click.INT, default=1, help='number of CPU cores to use')
@click.option('--seed', type=click.INT, default=0, help='random seed')
def binary(**kwargs):
    '''compute overlap/mutual exclusivity between features.

    This only considers presence and absence of features in the samples.'''
    compute_cooccur_workflow(method='binary', **kwargs)


@main.command()
@_common_parameters
@click.option('--clip', type=click.FLOAT, default=2, help='Clip (limit) the lower bound of feature values')
@click.option('--perms', type=click.INT, default=1000, help='the number of permutations to compute p-value')
@click.option('-c', '--cpus', type=click.INT, default=1, help='number of CPU cores to use')
@click.option('--seed', type=click.INT, default=0, help='random seed')
def logratio(**kwargs):
    '''compute correlation between features using log ratio.

    This is robust to compositionality effects.'''
    compute_cooccur_workflow(method='logratio', **kwargs)


@main.command()
@_common_parameters
@click.option('--clip', type=click.FLOAT, default=2, help='Clip (limit) the lower bound of feature values')
@click.option('--nonzero', is_flag=True, help='discard the values that are below clip in both features.')
def rank(**kwargs):
    '''compute non-parametric correlation between features.

    This ranks features within each sample first and then compute correlation between features'''
    compute_cooccur_workflow(method='rank', **kwargs)


@main.command()
@click.option('-s', '--stat-file', type=click.Path(exists=True), required=True, help='input file')
@click.option('-p', '--pval-file', type=click.Path(exists=True), required=True, help='input file')
@click.option('-o', '--output', type=click.File('w'), help='output file')
@click.option('-c', '--cutoff', type=click.FLOAT, default=0.01, help='p-value cutoff')
def convert(stat_file, pval_file, output, cutoff):
    '''convert file format to edge file format.'''
    if output is None:
        output = os.path.splitext(stat_file)[0] + '.tsv'
    pval = pd.read_csv(pval_file, sep='\t', index_col=0)
    stat = pd.read_csv(stat_file, sep='\t', index_col=0)
    df = convert_format(stat, pval, cutoff)
    df.to_csv(output, sep='\t', index=False)


def compute_cooccur_workflow(
        input, output,
        format='biom', method='binary',
        cpus=1, progress=True, force=False, verbose=1,
        **kwargs):
    '''
    Parameters
    ----------
    input : str
        input file path
    output : str
        output file prefix
    format : str
        input file format
    method : str
        co-occur method
    cpus : int
        CPU cores
    progress : bool
        whether to show progress bar
    force : bool
        whether to overwrite if output files exist
    verbose : int
        verbosity level
    kwargs : dict
        keyword argument passing to co-occur method

    Returns
    -------
    None
    '''
    levels = ['WARNING', 'INFO', 'DEBUG']
    n = len(levels)
    if verbose >= n:
        verbose = n - 1
    logger = getLogger()
    logger.setLevel(levels[verbose])

    o1 = output + '.stat'
    o2 = output + '.pval'
    if not force and os.path.exists(o1):
        raise FileExistsError('%s already exists' % output)
    if format == 'biom':
        from biom import load_table
        table = load_table(input)
        data = table.matrix_data.transpose()
        ids = table.ids(axis='observation')
    elif format == 'tsv':
        df = pd.read_csv(input, sep='\t', index_col=0, comment='#')
        data = df.values.T
        ids = df.index
    logger.debug('run method: %s' % method)
    stat, pval = compute_cooccur(data, method, cpus, progress, **kwargs)
    stat_df = pd.DataFrame(stat, columns=ids, index=ids)
    pval_df = pd.DataFrame(pval, columns=ids, index=ids)
    stat_df.to_csv(o1, sep='\t')
    pval_df.to_csv(o2, sep='\t')
    return stat_df, pval_df


if __name__ == "__main__":
    sys.exit(main())  # pragma: no cover


shenshang/util.py
from logging import getLogger

import numpy as np
import pandas as pd


def set_log_level(level):
    '''Set the log level.

    You can see the logging levels at:
    https://docs.python.org/3.5/library/logging.html#levels

    Parameters
    ----------
    level : int or str
        10 for debug, 20 for info, 30 for warn, etc.
        It is passing to :func:`logging.Logger.setLevel`

    '''
    logger = getLogger(__package__)
    logger.setLevel(level)


def argsort_arrays(arrs, ascending=True):
    '''argsort the input 1-D arrays sequentially.

    This function keeps the corresponding associations across input arrays.

    Parameters
    ----------
    arrs : a list of 1-D arrays
        positional arguments to accept a single or multiple 1-D arrays
    ascending : bool or a list of bool
        ascending or descending for each array

    Returns
    -------
    index_array : ndarray, int
        Array of indices that sort the input arrays in sequential order.
    '''
    idx = np.arange(arrs[0].size)
    if isinstance(ascending, bool):
        ascending = [ascending] * len(arrs)
    for x, a in zip(reversed(arrs), reversed(ascending)):
        idx2 = np.argsort(x[idx], kind='mergesort')
        idx = idx[idx2]
        if not a:
            idx = idx[::-1]
    return idx


def fetch_tri(square, lower=True):
    '''Return values in lower (or upper triangle) and their indices in the
square matrix.

    Parameters
    ----------
    square : 2-D square array

    lower : bool
        whether to fetch values in lower triangle or upper triangle

    Returns
    -------
    tuple of 3 arrays of the same length
        - values
        - row indices in the square matrix
        - column indices in the square matrix
    '''
    n = square.shape[0]
    if lower:
        idx = np.tril_indices(n, -1)
    else:
        idx = np.triu_indices(n, 1)
    v = square[idx]
    return v, idx[0], idx[1]


def convert_format(stat, pval, cutoff=0.01,
                   columns=('feature1', 'feature2', 'stat', 'pval', 'type')):
    '''Convert matrix to edge file format.

    The edge file format has each co-occurrence in a row. The first 2
    columns are the IDs of the the feature pairs.  The next columns
    are the statistic, p-value, and correlation type (negative or
    positive) for each feature pairs.

    The edge file can be provided to Cytoscape for network analysis.

    Parameters
    ----------
    stat, pval : pd.DataFrame
        square matrices of statistics and p-values between features on index and column
    cutoff : float
        p-value cutoff. All pairs with p-values in the interval [0, cutoff) will be kept
    columns : iterable of strings
        the column names for the return pd.DataFrame

    Returns
    -------
    pd.DataFrame
        1st column: feature 1
        2nd column: feature 2
        3rd column: statistic between features 1 and 2
        4th column: p-value between features 1 and 2
        5th column: '+' for positive correlation; '-' for negative correlation
    '''
    l = []
    ids = pval.index
    for lower, t in [(False, '+'), (True, '-')]:
        for p, i, j in zip(*fetch_tri(pval.values, lower=lower)):
            # stat is only located in the upper triangle
            if lower:
                s = stat.values[j, i]
            else:
                s = stat.values[i, j]
            # use `<` sign not only selects significant values, but also
            # discards np.nan too.
            if p < cutoff:
                l.append([ids[i], ids[j], s, p, t])
    return pd.DataFrame(l, columns=columns)


def compare_matrices(pvals, sigs, cutoff=0.01):
    '''Find the items across p-value matrices that satisfy the sigs.

    Parameters
    ----------
    pvals : list of square pd.DataFrame
        p-values from each methods
    sigs : tuple of bool
        filter to keep items that satisfy the significance patterns across pvals
    cutoff : float
        p-value cutoff

    Returns
    -------
    pd.DataFrame
        1st column: feature 1
        2nd column: feature 2
        3rd column: '+' for positive correlation; '-' for negative correlation
        rest columns: each one contains p-value for one method
    '''
    mask = np.ones_like(pvals[0], dtype=bool)
    assert len(pvals) == len(sigs)
    for pval, pat in zip(pvals, sigs):
        m = pval < cutoff
        # do NOT use `>` for False value of pat, because np.nan >
        # cutoff returns False while we want it to return True.
        # This is important for the p-values from rank method.
        if not pat:
            m = ~m
        mask = mask & m

    col1, col2 = np.where(mask)
    # a list of 1-D arrays
    p = [pval.values[mask.values] for pval in pvals]
    ids = pvals[0].index
    df = pd.DataFrame(
        dict(zip(range(3+len(p)),
                 ([ids[i] for i in col1],
                  [ids[i] for i in col2],
                  ['+' if i else '-' for i in col1 < col2],
                  *p))))
    return df


def combine(cutoff=0.05, **results):
    '''the results should be oriented the same.

    Parameters
    ----------
    results : dict
        keyword arguments. Key is the label text; value is the
        tuple/list of stat and pval that are square pd.DataFrame (eg
        `binary=[stat, pval], rank=[stat, pval]`)

    Returns
    -------
    '''
    dfs = []
    for i, name in enumerate(results, 1):
        stat, pval = results[name]
        df = convert_format(
            stat, pval, cutoff,
            columns=('feature1', 'feature2',
                     '{}_stat'.format(name),
                     '{}_pval'.format(name),
                     '{}_type'.format(name)))
        dfs.append(df.set_index(['feature1', 'feature2']))
    # import pdb; pdb.set_trace()
    return pd.concat(dfs, axis=1)


shenshang/visualize.py
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
from matplotlib.gridspec import GridSpec, GridSpecFromSubplotSpec

from .cooccur import cooccur_binary, cooccur_logratio
from .util import argsort_arrays


def plot_joint_scatter(x, y, cutoff=0, xlabel='x', ylabel='y',
                       ratio=3, fig=None, grid=None, equal_scale=True, **kwargs):
    '''Plot joint scatter and marginal histograms of x and y.

    Parameters
    ----------
    x, y : 1-D numeric arrays of the same size
        the abundance of the feature across all samples
    cutoff : numeric
        the threshold to determine the absence/presence of the feature
    xlabel, ylabel : str
        the label text
    ratio : numeric
        the width ratio of scatter plot over hist plot
    fig : matplotlib.figure.Figure or None (default)
        Figure object to draw the plot onto. None (default) to create a new figure
    grid : matplotlib.gridspec.GridSpec or None (default)
        GridSpect object to draw the plot onto
    equal_scale : bool
        whether to plot in same x-axis and y-axis view limits
    **kwargs : dict
        keyword arguments passing to :func:`matplotlib.axes.Axes.scatter`

    Returns
    -------
    matplotlib.figure.Figure, matplotlib.axes.Axes, matplotlib.axes.Axes, matplotlib.axes.Axes
    '''
    spec = dict(nrows=2, ncols=2,
                width_ratios=[ratio, 1],
                height_ratios=[1, ratio],
                hspace=0, wspace=0)
    if fig is None:
        fig = plt.figure()
        gs = GridSpec(**spec)
    else:
        if grid is None:
            raise ValueError('if `fig` is not None, you must provide value for `grid`.')
        gs = GridSpecFromSubplotSpec(subplot_spec=grid, **spec)

    ax_scatter = fig.add_subplot(gs[2])
    ax_histx = fig.add_subplot(gs[0], sharex=ax_scatter)
    ax_histy = fig.add_subplot(gs[3], sharey=ax_scatter)

    ax_scatter.scatter(x=x, y=y, **kwargs)
    if equal_scale:
        ymin, ymax = ax_scatter.get_ylim()
        xmin, xmax = ax_scatter.get_xlim()
        l, u = min(xmin, ymin), max(xmax, ymax)
        ax_scatter.set(xlim=(l, u), ylim=(l, u))
    ax_scatter.set(xlabel=xlabel, ylabel=ylabel)
    ax_scatter.axvline(cutoff, color='gray', linewidth=1, alpha=0.5)
    ax_scatter.axhline(cutoff, color='gray', linewidth=1, alpha=0.5)

    # ax[0].set_xscale('log')
    # ax[0].set_yscale('log')
    # ax[0].set_xscale('symlog', linthreshx=20)
    # ax[0].set_yscale('symlog', linthreshx=20)
    ax_histx.hist(x, bins=100)
    ax_histx.xaxis.set_visible(False)
    ax_histx.set_ylabel('count')

    ax_histy.hist(y, orientation='horizontal', bins=100)
    ax_histy.yaxis.set_visible(False)
    ax_histy.set_xlabel('count')

    return fig, ax_scatter, ax_histx, ax_histy


def plot(x, y, cutoff, psudo, perms, seed=None,
         xlabel='x', ylabel='y',
         colors='black', title='', ratio=3, height=4, **kwargs):
    '''Plot histogram of distribution of overlap or exclusivity for the permutated features.

    Parameters
    ----------
    x, y : 1-D numeric arrays of the same size
        the abundance of the feature across all samples
    psudo : numeric
        psudo count to add both nominator and denominator for overlap
        compute. The choice of psudo count value should depend on the
        sample size, i.e. the size of `x`.
    cutoff : numeric
        the threshold to determine the absence/presence of the feature
    perms : int
        the number of permutations to do for y.
    seed : int or None (default)
        random seed for random permutation.
    xlabel, ylabel : str
        the label text
    colors : str
        color used for the plot elements
    title : str
        figure title
    ratio : numeric
        the width ratio of hist plot over scatter plot
    height : numeric
        the height of figure
    **kwargs :
        keyword arguments passing to :func:`matplotlib.axes.Axes.scatter`

    Returns
    -------
    matplotlib.figure.Figure
        The Figure object containing the plot
    '''
    v1, p1_pos, p1_neg, dist1 = cooccur_binary(x, y, cutoff=cutoff, psudo=psudo, perms=perms, seed=seed)
    v2, p2_pos, p2_neg, dist2 = cooccur_logratio(x, y, clip=cutoff, perms=perms, seed=seed)

    outer = GridSpec(1, 3, width_ratios=[ratio+1, ratio, ratio])
    gs2 = GridSpecFromSubplotSpec(2, 1, subplot_spec=outer[1], height_ratios=[1, ratio], hspace=0)
    gs3 = GridSpecFromSubplotSpec(2, 1, subplot_spec=outer[2], height_ratios=[1, ratio], hspace=0)

    fig = plt.figure(figsize=(height / (ratio+1) * (1+ratio*3), height))

    if not title:
        title = 'x:{} y:{}'.format(np.sum(x>=cutoff), np.sum(y>=cutoff))
    fig.suptitle(title)

    plot_joint_scatter(x, y, xlabel=xlabel, ylabel=ylabel,
                       ratio=ratio, fig=fig, grid=outer[0],
                       c=colors, s=9, alpha=0.9, **kwargs)

    ax_hist_binary = fig.add_subplot(gs2[1])
    ax_hist_logratio = fig.add_subplot(gs3[1])

    ax_hist_binary.hist(dist1, bins=50)
    ax_hist_binary.set_xlabel('binary')
    ax_hist_binary.set_title('value: %.2f\npval+: %.2f\npval-: %.2f' % (v1, p1_pos, p1_neg))
    ax_hist_binary.axvline(v1, color='red')

    ax_hist_logratio.hist(dist2, bins=50)
    ax_hist_logratio.set_xlabel('logratio')
    ax_hist_logratio.set_title('value: %.2f\npval+: %.2f\npval-: %.2f' % (v2, p2_pos, p2_neg))
    ax_hist_logratio.axvline(v2, color='red')

    fig.tight_layout()

    return fig


def plot_logratio(x, y, clip=1, perms=50, seed=None, color='red', ax=None, **kwargs):
    '''Plot logratio of feature y and feature x in sorted order.

    Parameters
    ----------
    x, y : 1-D numeric arrays of the same size
        the abundance of the feature across all samples.
    clip : numeric
        the threshold to clip the feature values on the lower bound.
    perms : int
        the number of permutations to do for y.
    seed : int or None (default)
        random seed for random permutation.
    color : matplotlib color, optional
        facecolor of the shades.
    ax : matplotlib.axes.Axes, optional
        Axes object to draw the plot onto, otherwise uses the current Axes.
    kwargs : dict
        keyword arguments passing to the :func:`matplotlib.axes.Axes.set` function


    Returns
    -------
    matplotlib.axes.Axes
        The Axes object containing the plot.
    '''
    if ax is None:
        fig, ax = plt.subplots()
    x, y = np.copy(x), np.copy(y)
    x[x < clip] = clip
    y[y < clip] = clip
    n = x.size
    logratio = np.log(y/x)
    sorted_logratio = np.sort(logratio)
    ax.plot(sorted_logratio, color=color, linewidth=3, alpha=0.7)
    ax.fill_between(np.arange(n), sorted_logratio, facecolor=color, alpha=0.2)
    ax.set(ylabel='logratio log(y/x)', **kwargs)
    if seed is not None:
        np.random.seed(seed)
    for _ in range(perms):
        logratio = np.log(np.random.permutation(y)/x)
        sorted_logratio = np.sort(logratio)
        ax.plot(sorted_logratio, color='grey', linewidth=0.5, alpha=1)
        #ax.fill_between(np.arange(n), sorted_logratio, facecolor='gainsboro', alpha=0.1)
    return ax


def plot_sorted_bars(x, y, ascending=(True, False),
                     xlabel='x', ylabel='y', colors=None, ax=None, **kwargs):
    '''Plot the abundance of feature x and y in sorted order.

    Parameters
    ----------
    x, y : 1-D numeric arrays of the same size
        the abundance of the feature across all samples.
    xlabel, ylabel : str
        the label text
    colors : tuple of (str, str) or None (optional)
        The colors of the bar faces.
    ax : matplotlib.axes.Axes, optional
        Axes object to draw the plot onto, otherwise uses the current Axes.
    **kwargs :
        keyword arguments passing to :func:`matplotlib.axes.Axes.bar`

    Returns
    -------
    matplotlib.axes.Axes
        The Axes object containing the plot.
    '''
    if ax is None:
        fig, ax = plt.subplots()
    if colors is None:
        colors = plt.get_cmap('Set1').colors[:2]

    idx = argsort_arrays((x, y), ascending)
    ind = np.arange(x.size)
    ax.bar(ind, x[idx], color=colors[0], label=xlabel, **kwargs)
    ax.axhline(0, linewidth=2, color='black')
    ax.bar(ind, -y[idx], color=colors[1], label=ylabel, **kwargs)
    ax.set_xlabel('samples')
    ax.set_ylabel('abundance')
    ax.legend()
    return ax


def plot_sorted_shades(x, y, ascending=(True, False),
                       xlabel='x', ylabel='y', colors=None, ax=None, **kwargs):
    '''Sort x and y in the opposite direction and plot.

    Parameters
    ----------
    x, y : 1-D numeric arrays of the same size
        the abundance of the feature across all samples.
    xlabel, ylabel : str
        the label text
    colors : tuple of (str, str) or None (optional)
        The colors filling in the shades.
    ax : matplotlib.axes.Axes, optional
        Axes object to draw the plot onto, otherwise uses the current Axes.
    kwargs : dict
        keyword arguments passing to the :func:`matplotlib.axes.Axes.set` function


    Returns
    -------
    matplotlib.axes.Axes
        The Axes object containing the plot.
    '''
    if ax is None:
        fig, ax = plt.subplots()
    if colors is None:
        colors = plt.get_cmap('Set1').colors[:2]

    ind = np.arange(x.size)
    idx = argsort_arrays((x, y), ascending)
    x, y = x[idx], y[idx]
    ax.fill_between(ind, x, color=colors[0], alpha=0.3)
    ax.plot(ind, x, color=colors[0], alpha=0.6, label=xlabel)
    ax.fill_between(ind, y, color=colors[1], alpha=0.3)
    ax.plot(ind, y, color=colors[1], alpha=0.6, label=ylabel)
    ax.set(xlabel='samples', ylabel='abundance', **kwargs)
    ax.legend()
    return ax


def plot_confusion_matrix(x, y, cutoff,
                          xlabel='x', ylabel='y', ratio=True, mask=True, cmap=plt.cm.Blues,
                          ax=None, **kwargs):
    '''Plot confusion matrix of absence-presence for x and y.

    Parameters
    ----------
    x, y : 1-D numeric arrays of the same size
        the abundance of the feature across all samples
    cutoff : numeric
        the threshold to determine the absence/presence of the feature
    xlabel, ylabel : str
        the label text
    ratio : bool
        whether to plot the ratios (the counts in confusion matrix / the counts
        of features present in both samples) or the counts.
    mask : bool
        whether to mask (not show) the absence-absence cell in the heatmap
    ax : matplotlib.axes.Axes or None (default), optional
        Axes object to draw the plot onto; otherwise uses the current Axes
    cmap : string, optional
        matplotlib qualitative colormap
    kwargs : dict
        keyword arguments passing to the :func:`matplotlib.axes.Axes.set` function

    Returns
    -------
    matplotlib.axes.Axes
        The Axes object containing the plot
    '''
    x_bool = np.asarray(x) >= cutoff
    y_bool = np.asarray(y) >= cutoff
    pp = np.sum(x_bool & y_bool) # present in both x and y
    aa = x_bool.size - np.sum(x_bool | y_bool) # absent in both x and y
    pa = np.sum(x_bool) - pp # present in x, absent in y
    ap = np.sum(y_bool) - pp # present in y, absent in x
    # mask out absence-absence because we are not interested in it.
    data = np.ma.array([[aa, pa], [ap, pp]],
                       mask=[[mask, 0], [0, 0]])
    if ratio and pp != 0:
        data = data / pp
        fmt = '.2f'
    else:
        fmt = 'd'
    if ax is None:
        fig, ax = plt.subplots()
    im = ax.imshow(data, interpolation='nearest', cmap=cmap)
    ax.figure.colorbar(im, ax=ax)
    ax.patch.set(hatch='x', edgecolor='lightgray')
    ax.set(xticks=[0, 1], yticks=[0, 1],
           xticklabels=['Absent', 'Present'],
           yticklabels=['Absent', 'Present'],
           xlabel=xlabel,
           ylabel=ylabel,
           **kwargs)
    thresh = (data.max() - data.min()) / 2. + data.min()
    for i, j in zip(*np.where(~data.mask)):
        ax.text(i, j, format(data[j, i], fmt),
                ha="center", va="center",
                color="white" if data[j, i] > thresh else "black")
    return ax
